pkg/controllers/monitoring-stack/components.go

```diff
@@ -0,0 +1,343 @@
+					Type:    "prometheus",
+					Access:  "proxy",
+					Url:     prometheusUrl,
+					Version: 1,
+				},
+			},
+		},
+	}
+}
+
+func newRole(ms *stack.MonitoringStack, rbacResourceName string, rbacVerbs []string) *rbacv1.Role {
+	return &rbacv1.Role{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "rbac.authorization.k8s.io/v1",
+			Kind:       "Role",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rbacResourceName,
+			Namespace: ms.Namespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"services", "endpoints", "pods"},
+				Verbs:     rbacVerbs,
+			},
+			{
+				APIGroups: []string{"extensions", "networking.k8s.io"},
+				Resources: []string{"ingresses"},
+				Verbs:     rbacVerbs,
+			},
+		},
+	}
+}
+
+func newServiceAccount(ms *stack.MonitoringStack) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "ServiceAccount",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ms.Name + "-prometheus",
+			Namespace: ms.Namespace,
+		},
+	}
+}
+
+func newPrometheus(
+	ms *stack.MonitoringStack,
+	rbacResourceName string,
+	instanceSelectorKey string,
+	instanceSelectorValue string,
+) *monv1.Prometheus {
+	prometheusSelector := ms.Spec.ResourceSelector
+	if prometheusSelector == nil {
+		prometheusSelector = &metav1.LabelSelector{}
+	}
+	prometheus := &monv1.Prometheus{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Prometheus",
+			APIVersion: "monitoring.coreos.com/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ms.Name,
+			Namespace: ms.Namespace,
+			Labels:    prometheusLabels(ms.Name, instanceSelectorKey, instanceSelectorValue),
+		},
+
+		Spec: monv1.PrometheusSpec{
+			Image: nil,
+			// Prometheus does not use an Enum for LogLevel, so need to convert to string
+			LogLevel: string(ms.Spec.LogLevel),
+
+			Retention: ms.Spec.Retention,
+			Resources: ms.Spec.Resources,
+
+			ServiceAccountName: rbacResourceName,
+
+			ServiceMonitorSelector:          prometheusSelector,
+			ServiceMonitorNamespaceSelector: nil,
+			PodMonitorSelector:              prometheusSelector,
+			PodMonitorNamespaceSelector:     nil,
+		},
+	}
+	return prometheus
+}
+
+func newRoleBinding(ms *stack.MonitoringStack, rbacResourceName string) *rbacv1.RoleBinding {
+	roleBinding := &rbacv1.RoleBinding{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "rbac.authorization.k8s.io/v1",
+			Kind:       "RoleBinding",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rbacResourceName,
+			Namespace: ms.Namespace,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				APIGroup:  "",
+				Kind:      "ServiceAccount",
+				Name:      rbacResourceName,
+				Namespace: ms.Namespace,
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "Role",
+			Name:     rbacResourceName,
+		},
+	}
+	return roleBinding
+}
+
+func newServiceMonitor(ms *stack.MonitoringStack) *monv1.ServiceMonitor {
```

Comment:
This does not work when a monitoring stack has label selectors because the service monitor will not get picked up. Should we use `additionalScrapeConfigs` in the `PrometheusSpec` to make sure that setting selectors will not interfere with self-scraping?

https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#prometheusspec